src/DynExpManager/Instruments/NP_Conex_CC.cpp

```diff
@@ -0,0 +1,390 @@
+// This file is part of DynExp.
+
+#include "stdafx.h"
+#include "NP_Conex_CC.h"
+
+namespace DynExpInstr
+{
+	void NP_Conex_CC_Tasks::InitTask::InitFuncImpl(dispatch_tag<PositionerStageTasks::InitTask>, DynExp::InstrumentInstance& Instance)
+	{
+		{
+			auto InstrParams = DynExp::dynamic_Params_cast<NP_Conex_CC>(Instance.ParamsGetter());
+			auto InstrData = DynExp::dynamic_InstrumentData_cast<NP_Conex_CC>(Instance.InstrumentDataGetter());
+
+			Instance.LockObject(InstrParams->HardwareAdapter, InstrData->HardwareAdapter);
+			InstrData->HardwareAdapter->Clear();
+
+			// Define and go to home (this includes resetting the controller):
+			// To set the current position as home position, the stage has to be in the CONFIGURATION state. This state can only be reached from the NOT REFERENCED state. 
+			*InstrData->HardwareAdapter << Util::ToStr(InstrParams->ConexAddress.Get()) + "RS"; // Go to NOT REFERENCED state.
+			std::this_thread::sleep_for(std::chrono::milliseconds(500)); // This takes 500 ms.
+			*InstrData->HardwareAdapter << Util::ToStr(InstrParams->ConexAddress.Get()) + "PW1"; // Go to CONFIGURATION state.
+			*InstrData->HardwareAdapter << Util::ToStr(InstrParams->ConexAddress.Get()) + "HT1"; // Set current position to home position.
+			*InstrData->HardwareAdapter << Util::ToStr(InstrParams->ConexAddress.Get()) + "PW0"; // Go to NOT REFERENCED state.
+			std::this_thread::sleep_for(std::chrono::seconds(3)); // This takes 3 s.
+			*InstrData->HardwareAdapter << Util::ToStr(InstrParams->ConexAddress.Get()) + "OR"; // Go to READY state.
+
+		}
+
+		InitFuncImpl(dispatch_tag<InitTask>(), Instance);
+	}
+
+	void NP_Conex_CC_Tasks::ExitTask::ExitFuncImpl(dispatch_tag<PositionerStageTasks::ExitTask>, DynExp::InstrumentInstance& Instance)
+	{
+		ExitFuncImpl(dispatch_tag<ExitTask>(), Instance);
+
+		auto InstrParams = DynExp::dynamic_Params_cast<NP_Conex_CC>(Instance.ParamsGetter());
+		auto InstrData = DynExp::dynamic_InstrumentData_cast<NP_Conex_CC>(Instance.InstrumentDataGetter());
+
+		try
+		{
+			// Abort motion.
+			*InstrData->HardwareAdapter << Util::ToStr(InstrParams->ConexAddress.Get()) + "ST";
+		}
+		catch (...)
+		{
+			// Swallow any exception which might arise from HardwareAdapter->operator<<() since a failure
+			// of this function is not considered a severe error.
+		}
+
+		Instance.UnlockObject(InstrData->HardwareAdapter);
+	}
+
+	void NP_Conex_CC_Tasks::UpdateTask::UpdateFuncImpl(dispatch_tag<PositionerStageTasks::UpdateTask>, DynExp::InstrumentInstance& Instance)
+	{
+		try
+		{
+			auto InstrParams = DynExp::dynamic_Params_cast<NP_Conex_CC>(Instance.ParamsGetter());
```

Comment:
To avoid locking the params structure in every operation (which is expensive), store the channel in the data structure in the init task and take it from there. See implementation of SmarAct stage.